File: backend/models/watchlist.go
package models

import (
	"time"
)

// WatchList represents a user's watch list
type WatchList struct {
	ID           string    `json:"id" db:"id"`
	UserID       string    `json:"user_id" db:"user_id"`
	Name         string    `json:"name" db:"name"`
	Description  *string   `json:"description" db:"description"`
	IsDefault    bool      `json:"is_default" db:"is_default"`
	DisplayOrder int       `json:"display_order" db:"display_order"`
	IsPublic     bool      `json:"is_public" db:"is_public"`
	PublicSlug   *string   `json:"public_slug,omitempty" db:"public_slug"`
	CreatedAt    time.Time `json:"created_at" db:"created_at"`
	UpdatedAt    time.Time `json:"updated_at" db:"updated_at"`
}

// WatchListItem represents a ticker in a watch list
type WatchListItem struct {
	ID              string    `json:"id" db:"id"`
	WatchListID     string    `json:"watch_list_id" db:"watch_list_id"`
	Symbol          string    `json:"symbol" db:"symbol"`
	Notes           *string   `json:"notes" db:"notes"`
	Tags            []string  `json:"tags" db:"tags"`
	TargetBuyPrice  *float64  `json:"target_buy_price" db:"target_buy_price"`
	TargetSellPrice *float64  `json:"target_sell_price" db:"target_sell_price"`
	AddedAt         time.Time `json:"added_at" db:"added_at"`
	DisplayOrder    int       `json:"display_order" db:"display_order"`
}

// WatchListItemWithData includes ticker data and real-time price
type WatchListItemWithData struct {
	WatchListItem
	// Ticker info from tickers table
	Name      string  `json:"name"`
	Exchange  string  `json:"exchange"`
	AssetType string  `json:"asset_type"`
	LogoURL   *string `json:"logo_url"`
	// Real-time price data
	CurrentPrice   *float64 `json:"current_price"`
	PriceChange    *float64 `json:"price_change"`
	PriceChangePct *float64 `json:"price_change_pct"`
	Volume         *int64   `json:"volume"`
	MarketCap      *float64 `json:"market_cap"`
	PrevClose      *float64 `json:"prev_close"`
	// Reddit data (from reddit_heatmap_daily table)
	RedditRank       *int     `json:"reddit_rank,omitempty"`        // Current Reddit rank (1 = #1 trending)
	RedditMentions   *int     `json:"reddit_mentions,omitempty"`    // Total mentions
	RedditPopularity *float64 `json:"reddit_popularity,omitempty"`  // Popularity score (0-100)
	RedditTrend      *string  `json:"reddit_trend,omitempty"`       // "rising", "falling", "stable"
	RedditRankChange *int     `json:"reddit_rank_change,omitempty"` // Rank change vs 24h ago
}

// WatchListItemDetail extends WatchListItemWithData with screener_data fields
// (IC Score, fundamentals, valuation ratios, and alert count).
// Pointer fields serialize as null (not omitted) so the frontend gets a stable schema.
type WatchListItemDetail struct {
	WatchListItemWithData

	// IC Score (from screener_data materialized view)
	ICScore               *float64 `json:"ic_score"`
	ICRating              *string  `json:"ic_rating"`
	ValueScore            *float64 `json:"value_score"`
	GrowthScore           *float64 `json:"growth_score"`
	ProfitabilityScore    *float64 `json:"profitability_score"`
	FinancialHealthScore  *float64 `json:"financial_health_score"`
	MomentumScore         *float64 `json:"momentum_score"`
	AnalystConsensusScore *float64 `json:"analyst_consensus_score"`
	InsiderActivityScore  *float64 `json:"insider_activity_score"`
	InstitutionalScore    *float64 `json:"institutional_score"`
	NewsSentimentScore    *float64 `json:"news_sentiment_score"`
	TechnicalScore        *float64 `json:"technical_score"`
	SectorPercentile      *float64 `json:"sector_percentile"`
	LifecycleStage        *string  `json:"lifecycle_stage"`

	// Fundamentals (from screener_data)
	PERatio         *float64 `json:"pe_ratio"`
	PBRatio         *float64 `json:"pb_ratio"`
	PSRatio         *float64 `json:"ps_ratio"`
	ROE             *float64 `json:"roe"`
	ROA             *float64 `json:"roa"`
	GrossMargin     *float64 `json:"gross_margin"`
	OperatingMargin *float64 `json:"operating_margin"`
	NetMargin       *float64 `json:"net_margin"`
	DebtToEquity    *float64 `json:"debt_to_equity"`
	CurrentRatio    *float64 `json:"current_ratio"`
	RevenueGrowth   *float64 `json:"revenue_growth"`
	EPSGrowth       *float64 `json:"eps_growth"`
	DividendYield   *float64 `json:"dividend_yield"`
	PayoutRatio     *float64 `json:"payout_ratio"`

	// Alert metadata
	AlertCount int        `json:"alert_count"`
	Alert      *AlertRule `json:"alert,omitempty"` // The single alert for this item (populated when include_alerts=true)
}

// WatchListWithItems includes the watch list, all items with detail, and summary metrics
type WatchListWithItems struct {
	WatchList
	ItemCount         int                      `json:"item_count"`
	Items             []WatchListItemDetail    `json:"items"`
	Summary           *WatchListSummaryMetrics `json:"summary,omitempty"`
	AlertsFetchFailed bool                     `json:"alerts_fetch_failed,omitempty"` // True when alert enrichment fails; lets the UI show degraded state instead of "no alerts"
}

// WatchListSummaryMetrics provides aggregate stats across all items in a watchlist
type WatchListSummaryMetrics struct {
	TotalTickers        int      `json:"total_tickers"`
	AvgICScore          *float64 `json:"avg_ic_score,omitempty"`
	AvgDayChangePct     *float64 `json:"avg_day_change_pct,omitempty"`
	AvgDividendYield    *float64 `json:"avg_dividend_yield,omitempty"`
	RedditTrendingCount int      `json:"reddit_trending_count"`
}

// WatchListSummary is a lightweight version for listing watch lists
type WatchListSummary struct {
	ID          string    `json:"id"`
	Name        string    `json:"name"`
	Description *string   `json:"description"`
	IsDefault   bool      `json:"is_default"`
	ItemCount   int       `json:"item_count"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// Request/Response DTOs

// CreateWatchListRequest for creating a new watch list
type CreateWatchListRequest struct {
	Name        string  `json:"name" binding:"required,min=1,max=255"`
	Description *string `json:"description" binding:"omitempty,max=5000"`
}

// UpdateWatchListRequest for updating watch list metadata
type UpdateWatchListRequest struct {
	Name        string  `json:"name" binding:"min=1,max=255"`
	Description *string `json:"description" binding:"omitempty,max=5000"`
}

// AddTickerRequest for adding a ticker to watch list
type AddTickerRequest struct {
	Symbol          string   `json:"symbol" binding:"required,min=1,max=20"`
	Notes           *string  `json:"notes" binding:"omitempty,max=10000"`
	Tags            []string `json:"tags" binding:"max=50,dive,max=100"`
	TargetBuyPrice  *float64 `json:"target_buy_price" binding:"omitempty,gte=0"`
	TargetSellPrice *float64 `json:"target_sell_price" binding:"omitempty,gte=0"`
}

// UpdateTickerRequest for updating ticker metadata
type UpdateTickerRequest struct {
	Notes           *string  `json:"notes" binding:"omitempty,max=10000"`
	Tags            []string `json:"tags" binding:"max=50,dive,max=100"`
	TargetBuyPrice  *float64 `json:"target_buy_price" binding:"omitempty,gte=0"`
	TargetSellPrice *float64 `json:"target_sell_price" binding:"omitempty,gte=0"`
}

// BulkAddTickersRequest for CSV import
type BulkAddTickersRequest struct {
	Symbols []string `json:"symbols" binding:"required,min=1,max=500,dive,min=1,max=20"`
}

// ReorderItemsRequest for updating display order
type ReorderItemsRequest struct {
	ItemOrders []ItemOrder `json:"item_orders" binding:"required,min=1,max=500"`
}

type ItemOrder struct {
	ItemID       string `json:"item_id" binding:"required,max=100"`
	DisplayOrder int    `json:"display_order" binding:"min=0,max=10000"`
}

// TagWithCount represents a tag name and the number of watchlist items using it.
type TagWithCount struct {
	Name  string `json:"name" db:"name"`
	Count int    `json:"count" db:"count"`
}

File: lib/api/watchlist.ts
import { apiClient } from './client';
import { watchlists } from './routes';

export interface WatchList {
  id: string;
  user_id: string;
  name: string;
  description?: string;
  is_default: boolean;
  item_count: number;
  created_at: string;
  updated_at: string;
}

export interface WatchListItem {
  id: string;
  watch_list_id: string;
  symbol: string;
  notes?: string;
  tags: string[];
  target_buy_price?: number;
  target_sell_price?: number;
  added_at: string;
  display_order: number;
  // Ticker data
  name: string;
  exchange: string;
  asset_type: string;
  logo_url?: string;
  // Real-time price data
  current_price?: number;
  price_change?: number;
  price_change_pct?: number;
  volume?: number;
  market_cap?: number;
  prev_close?: number;
  // Reddit data
  reddit_rank?: number;
  reddit_mentions?: number;
  reddit_popularity?: number;
  reddit_trend?: string;
  reddit_rank_change?: number;
  // IC Score (from screener_data — always present, null when unavailable)
  ic_score: number | null;
  ic_rating: string | null;
  value_score: number | null;
  growth_score: number | null;
  profitability_score: number | null;
  financial_health_score: number | null;
  momentum_score: number | null;
  analyst_consensus_score: number | null;
  insider_activity_score: number | null;
  institutional_score: number | null;
  news_sentiment_score: number | null;
  technical_score: number | null;
  sector_percentile: number | null;
  lifecycle_stage: string | null;
  // Fundamentals (from screener_data — always present, null when unavailable)
  pe_ratio: number | null;
  pb_ratio: number | null;
  ps_ratio: number | null;
  roe: number | null;
  roa: number | null;
  gross_margin: number | null;
  operating_margin: number | null;
  net_margin: number | null;
  debt_to_equity: number | null;
  current_ratio: number | null;
  revenue_growth: number | null;
  eps_growth: number | null;
  dividend_yield: number | null;
  payout_ratio: number | null;
  // Alert metadata
  alert_count: number;
}

export interface WatchListSummaryMetrics {
  total_tickers: number;
  avg_ic_score?: number;
  avg_day_change_pct?: number;
  avg_dividend_yield?: number;
  reddit_trending_count: number;
}

export interface WatchListWithItems {
  id: string;
  name: string;
  description?: string;
  is_default: boolean;
  item_count: number;
  items: WatchListItem[];
  summary?: WatchListSummaryMetrics;
  created_at: string;
  updated_at: string;
}

export const watchListAPI = {
  // Get all watch lists for user
  async getWatchLists(): Promise<{ watch_lists: WatchList[] }> {
    return apiClient.get(watchlists.list);
  },

  // Create new watch list
  async createWatchList(data: { name: string; description?: string }): Promise<WatchList> {
    return apiClient.post(watchlists.create, data);
  },

  // Get single watch list with items
  async getWatchList(id: string): Promise<WatchListWithItems> {
    return apiClient.get(watchlists.byId(id));
  },

  // Update watch list metadata
  async updateWatchList(id: string, data: { name: string; description?: string }): Promise<void> {
    return apiClient.put(watchlists.byId(id), data);
  },

  // Delete watch list
  async deleteWatchList(id: string): Promise<void> {
    return apiClient.delete(watchlists.byId(id));
  },

  // Add ticker to watch list
  async addTicker(
    watchListId: string,
    data: {
      symbol: string;
      notes?: string;
      tags?: string[];
      target_buy_price?: number;
      target_sell_price?: number;
    }
  ): Promise<WatchListItem> {
    return apiClient.post(watchlists.items.add(watchListId), data);
  },

  // Remove ticker from watch list
  async removeTicker(watchListId: string, symbol: string): Promise<void> {
    return apiClient.delete(watchlists.items.remove(watchListId, symbol));
  },

  // Update ticker metadata
  async updateTicker(
    watchListId: string,
    symbol: string,
    data: {
      notes?: string;
      tags?: string[];
      target_buy_price?: number;
      target_sell_price?: number;
    }
  ): Promise<WatchListItem> {
    return apiClient.put(watchlists.items.update(watchListId, symbol), data);
  },

  // Bulk add tickers
  async bulkAddTickers(
    watchListId: string,
    symbols: string[]
  ): Promise<{
    added: string[];
    failed: string[];
    total: number;
  }> {
    return apiClient.post(watchlists.bulk(watchListId), { symbols });
  },

  // Get all tags used across user's watchlists with usage counts, ordered by popularity
  async getUserTags(): Promise<{ tags: { name: string; count: number }[] }> {
    return apiClient.get(watchlists.tags);
  },

  // Reorder items
  async reorderItems(
    watchListId: string,
    itemOrders: Array<{ item_id: string; display_order: number }>
  ): Promise<void> {
    return apiClient.post(watchlists.reorder(watchListId), { item_orders: itemOrders });
  },
};
